File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章统计子矩阵.cpp
//统计子矩阵    lanqiaoOJ 2109

#include<bits/stdc++.h>
using namespace std;
int s[550][550];
int main(){
    int n,m,k; cin>>n>>m>>k;
    for(int i=1;i<=n;i++)
        for(int j=1;j<=m;j++) {
            int a;   scanf("%d",&a);
            s[i][j] = s[i-1][j] + a;         // s[i][j]：第j列上，0~i行数字的前缀和
        }
    long long ans=0;
    for(int i1=1;i1<=n;i1++)
        for(int i2=i1;i2<=n;i2++)
            for(int j1=1,j2=1,z=0;j2<=m;j2++){  //尺取法，滑动窗口[j1, j2]。移动指针j2
                z += s[i2][j2]-s[i1-1][j2];  //第j2列上，i1~i2的区间和。累加得到二维区间和
                while(z>k){                  //若区间和 > k，移动指针j1
                    z -= s[i2][j1]-s[i1-1][j1];
                    j1 += 1;
                }
                ans += j2-j1+1;            //若j1~j2的区间和 < k，那么这里面有j2-j1+1个满足
            }
    cout << ans;
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章统计子矩阵.py
#统计子矩阵   lanqiaoOJ 2109

n, m, k = map(int, input().split())
a = [[0] for i in range(n)]
a.insert(0,[0]*(m+1))
for i in range(1,n+1):                 #从a[1][1]开始，读矩阵
    a[i].extend(map(int, input().split()))
s = [[0]*(m+1) for i in range(n+1)]
for i in range(1,n+1):
    for j in range(1,m+1):
        s[i][j] = s[i-1][j] + a[i][j]
ans = 0
for i1 in range(1,n+1):
    for i2 in range(i1,n+1):
        j1=1; z=0
        for j2 in range(1,m+1):                                            
            z += s[i2][j2]-s[i1-1][j2]  
            while z>k:                     
                z -= s[i2][j1]-s[i1-1][j1]
                j1 += 1             
            ans += j2-j1+1            
print(ans)

